### PR TITLE
New: RuleTester supports grouping tests into scenarios

### PR DESCRIPTION
**Pull request by platinumazure**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Changed `lib/testers/rule-tester.js` so that valid and invalid tests can now be grouped into "scenarios". Basically, instead of passing an array of tests, you can pass an object. The keys of that object are scenario descriptions, and the values are either an array of tests or another object with sub-scenarios.

**Is there anything you'd like reviewers to focus on?**

RuleTester error handling/coverage is pretty light in general. Please let me know if I should try to cover more error cases.

Goal is for this to be fully backwards-compatible. Pretty sure I've achieved that or else tests would be failing throughout, but still.

### CODE REVIEW

**Comment by platinumazure** [comment]
I've created a branch on my fork to demonstrate how this would work: https://github.com/platinumazure/eslint/tree/eqeqeq-using-scenarios

And here's a snippet of what the output of the eqeqeq tests looks like on my branch:

eqeqeq-scenario-groups

File: tests/lib/testers/rule-tester.js [hunk 1]
```diff
@@ -636,4 +643,201 @@ describe("RuleTester", function() {
...
+            });
+        });
+
+        it("should throw if 'invalid' value is not an array or object", function() {
+            assert.throws(function() {
+                ruleTester.run("foo", emptyRule, {
+                    valid: [],
+                    invalid: "not an array or object"
+                });
+            });
```
Comment (platinumazure) [inline tests/lib/testers/rule-tester.js]: This probably needs an error message regex.

File: tests/lib/testers/rule-tester.js [hunk 2]
```diff
@@ -636,4 +643,201 @@ describe("RuleTester", function() {
...
+
+            assert.throws(function() {
+                ruleTester.run("foo", emptyRule, {
+                    valid: [],
+                    invalid() {}
+                });
+            });
```
Comment (platinumazure) [inline tests/lib/testers/rule-tester.js]: This probably needs an error message regex.

**Comment by platinumazure** [comment]
Not sure what happened with the builds.

**Comment by gyandeeps** [comment]
I also got the same errors: https://github.com/eslint/eslint/pull/7465

**Comment by not-an-aardvark** [comment]
Only the "PR" builds seem to be failing, and not the "push" builds, so maybe something broke on master.

**Comment by not-an-aardvark** [comment]
I can reproduce locally by doing a fresh `npm install`.

**Comment by gyandeeps** [comment]
I have found the cause. Its because of `mock-fs` 3.12.0 version which just got released like 10-15mins ago
It works fine with 3.11.0
Looking into what is gng on

**Comment by not-an-aardvark** [comment]
Should we lock the mock-fs version for now to fix the build, and then upgrade it later?

**Comment by gyandeeps** [comment]
Done: https://github.com/eslint/eslint/pull/7466

**Comment by not-an-aardvark** [comment]
@platinumazure I think the build should be fixed if you rebase this branch.

**Comment by platinumazure** [comment]
@gyandeeps @not-an-aardvark Thanks for looking into the issues and helping me get this to pass again!

**Comment by nzakas** [comment]
Is there an issue for this change? I'd like to know the background that led to this proposal. I ask because I'm not crazy about this idea - I'd rather rely on the unit testing framework (i.e. Mocha in our case) for having this level of granularity in tests. However, without knowing the problem this is supposed to address, I'm unsure of my position.

(Reminder: core changes are supposed to start with an issue. That way, we get a lot more eyes on them so we can figure out what sort of effects it will have on the ecosystem.)

**Comment by nzakas** [comment]
Just a heads up: we have moved to a new CLA checker on pull requests. Even if you've previously signed our CLA, we will need to you sign the new one. To do so, look at the status checks for **licence/cla** and click the "Details" link. Sorry for the inconvenience.

**Comment by platinumazure** [comment]
@nzakas Apologies, you are right, I should have created an issue first. Please see #7500. Sorry for the inconvenience.

**Comment by platinumazure** [comment]
Closing for now, as it is not clear this is needed. I will experiment with using `RuleTester.describe` followed by `RuleTester.run` and see how that looks for grouping related test cases together.

**Pull request closed without merging**